src/read.rs

```diff
@@ -0,0 +1,63 @@
+use std::io;
+use std::io::Write;
+use std::marker::PhantomData;
+use Event;
+
+pub struct ReadHtml<'a, I> where I: Iterator<Item=Event<'a>> + 'a {
+    events: I,
+    current: Vec<u8>,
+    phantom: PhantomData<&'a I>,
+}
+
+impl<'a, I> ReadHtml<'a, I> where I: Iterator<Item=Event<'a>> {
+    pub fn new<II>(events: II) -> ReadHtml<'a, I> where II: IntoIterator<IntoIter=I, Item=Event<'a>> {
+        ReadHtml {
+            events: events.into_iter(),
+            current: Vec::new(),
+            phantom: PhantomData,
+        }
+    }
+}
+
+impl<'a, I> io::Read for ReadHtml<'a, I> where I: Iterator<Item=Event<'a>> {
+    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
+        // TODO handle buf.len() == 0
```

Comment:
The case that is an issue is as you mentioned elsewhere if there are 0-length events. Then `len` will be 0 and it will return `Ok(0)` even though there's more events to read.

Could change the first if to a `while`, then if 0-length events do occur it will keep writing them till either it runs out of events to write or finds one that actually has some data.